src/wallet.cpp

```diff
@@ -1485,50 +1486,85 @@ bool CWallet::SelectCoins(int64_t nTargetValue, unsigned int nSpendTime, set<pai
+    std::set<pair<const CWalletTx*,unsigned int> >& setCoinsRet, bool fMiner) const
 {
+    int64_t BalanceToConsider = GetBalance();
+
+    // Check if we have a spendable balance
+    if (BalanceToConsider <= 0)
+    {
+        if (fMiner)
+        {
+            LOCK(MinerStatus.lock);
+            MinerStatus.Clear();
+            MinerStatus.ReasonNotStaking = _("No coins");
+        }
+
+        return false;
+    }
+    // Check if we have a balance after the reserve is applied to consider staking with
+    BalanceToConsider -= nReserveBalance;
+
+    if (BalanceToConsider <= 0)
+    {
+        if (fMiner)
+        {
+            LOCK(MinerStatus.lock);
+            MinerStatus.Clear();
+            MinerStatus.ReasonNotStaking = _("Entire balance reserved");
+        }
+
+        return false;
+    }
+
+    if (fDebug2 && fMiner)
+        LogPrintf("SelectCoinsForStaking: Balance considered for staking %.8f", BalanceToConsider / (double)COIN);
+
     vector<COutput> vCoins;
     AvailableCoinsForStaking(vCoins, nSpendTime);
 
-    setCoinsRet.clear();
-    nValueRet = 0;
+    // Check to see if any coins are available for staking
+    if (vCoins.empty())
+    {
+        if (fMiner)
+        {
+            LOCK(MinerStatus.lock);
+            MinerStatus.Clear();
+            MinerStatus.ReasonNotStaking = _("No mature coins");
+        }
 
-    int64_t nTargetValue = nTargetValueIn;
+        return false;
+    }
 
-    //if (GlobalCPUMiningCPID.cpid != "INVESTOR"  && msMiningErrors7 != "Probing coin age")
-    //{
-    //      nTargetValue = nTargetValueIn/2;
-    //}
+    // Iterate through the wallet of stakable utxos and return them to miner if we can stake with them
+    setCoinsRet.clear();
 
     for(const COutput& output : vCoins)
     {
         const CWalletTx *pcoin = output.tx;
         int i = output.i;
-
-        // Stop if we've chosen enough inputs
-        if (nValueRet >= nTargetValue)
-            break;
-
         int64_t n = pcoin->vout[i].nValue;
 
-        pair<int64_t,pair<const CWalletTx*,unsigned int> > coin = make_pair(n,make_pair(pcoin, i));
-
-        if (n >= nTargetValue)
+        // If the Spendable balance is more then utxo value it is classified as able to stake
+        if (BalanceToConsider >= n)
         {
-            // If input value is greater or equal to target then simply insert
-            //    it into the current subset and exit
-            setCoinsRet.insert(coin.second);
-            nValueRet += coin.first;
-            break;
-        }
-        else if (n < nTargetValue + CENT)
-        {
-            setCoinsRet.insert(coin.second);
-            nValueRet += coin.first;
+            if (fDebug2 && fMiner)
+                LogPrintf("SelectCoinsForStaking: UTXO=%s (BalanceToConsider=%.8f >= Value=%.8f)", pcoin->vout[i].GetHash().ToString(), BalanceToConsider / (double)COIN, n / (double)COIN);
+
+            setCoinsRet.insert(make_pair(pcoin, i));
```

Comment:
You should go ahead and put the std:shuffle here, if you can shuffle a set. (Not sure about that. If not, shuffle the vCoins above.) Using a uniform distribution random generator. Either way the set returned to the miner will be randomized. (Shuffling the vCoins may be less expensive than the set on a per element basis, but there will be more elements than the set if a reserve is set.)

I am wondering whether the dust check nValue/1250000 i.e. 1/80 GRC for a weight of zero belongs here or in the miner loop. Is it more proper to feed a UTXO that will have zero weight to the miner and let the miner take care of it or do it here?

Reply:
yes and yes. can do. just was working on the multisig consolidator ill focus on this in the morning

Reply:
Awesome stuff Paul!